File: matiasmasso/Mat.NET/DT2/DTOTaskResult.vb
Public Class DTOTaskResult
    Property ResultCod As DTOTask.ResultCods
    Property Msg As String
    Property Exceptions As List(Of Exception)

    Public Sub New()
        MyBase.New
        _Exceptions = New List(Of Exception)
    End Sub

    Shared Function Factory(oResultCod As DTOTask.ResultCods, Optional msg As String = "", Optional exs As List(Of Exception) = Nothing) As DTOTaskResult
        Dim retval As New DTOTaskResult
        retval.ResultCod = oResultCod
        retval.Msg = msg
        If exs IsNot Nothing Then
            retval.AddExceptions(exs)
        End If
        Return retval
    End Function

    Public Function Success() As Boolean
        Dim retval As Boolean
        Select Case _ResultCod
            Case DTOTask.ResultCods.Success, DTOTask.ResultCods.Empty
                retval = True
        End Select
        Return retval
    End Function

    Shared Function SuccessResult(Optional msg As String = "") As DTOTaskResult
        Dim retval As New DTOTaskResult
        retval.ResultCod = DTOTask.ResultCods.Success
        retval.Msg = msg
        Return retval
    End Function

    Shared Function FailResult(exs As List(Of System.Exception), Optional msg As String = "") As DTOTaskResult
        Dim retval As New DTOTaskResult
        With retval
            .ResultCod = DTOTask.ResultCods.Failed
            .AddExceptions(exs)
            .Msg = msg
        End With
        Return retval
    End Function

    Public Sub Fail()
        _ResultCod = DTOTask.ResultCods.Failed
    End Sub

    Public Sub Fail(stringFormatMsg As String, ParamArray stringFormatValues() As String)
        Fail()
        If stringFormatValues.Length > 0 Then
            _Msg = String.Format(stringFormatMsg, stringFormatValues)
        Else
            _Msg = stringFormatMsg
        End If
    End Sub

    Public Sub Fail(ex As System.Exception, stringFormatMsg As String, ParamArray stringFormatValues() As String)
        Fail(stringFormatMsg, stringFormatValues)
        AddException(ex)
    End Sub

    Public Sub Fail(exs As IEnumerable(Of System.Exception), stringFormatMsg As String, ParamArray stringFormatValues() As String)
        Fail(stringFormatMsg, stringFormatValues)
        AddExceptions(exs)
    End Sub

    Public Function SystemExceptions() As List(Of System.Exception)
        Dim retval As New List(Of System.Exception)
        For Each ex As Exception In _Exceptions
            retval.Add(New System.Exception(ex.Message))
        Next
        Return retval
    End Function




    Public Sub Succeed()
        _ResultCod = DTOTask.ResultCods.Success
    End Sub

    Public Sub Succeed(stringFormatMsg As String, ParamArray stringFormatValues() As String)
        Succeed()
        If stringFormatValues.Length > 0 Then
            _Msg = String.Format(stringFormatMsg, stringFormatValues)
        Else
            _Msg = stringFormatMsg
        End If
    End Sub

    Public Sub Empty()
        _ResultCod = DTOTask.ResultCods.Empty
    End Sub

    Public Sub Empty(stringFormatMsg As String, ParamArray stringFormatValues() As String)
        Empty()
        If stringFormatValues.Length > 0 Then
            _Msg = String.Format(stringFormatMsg, stringFormatValues)
        Else
            _Msg = stringFormatMsg
        End If
    End Sub

    Public Sub AddException(stringFormatMsg As String, ParamArray stringFormatValues() As String)
        Dim ex As Exception = Nothing
        If stringFormatValues.Length > 0 Then
            ex = New Exception(String.Format(stringFormatMsg, stringFormatValues))
        Else
            ex = New Exception(stringFormatMsg)
        End If
        _Exceptions.Add(ex)
        If _ResultCod = DTOTask.ResultCods.Success Then _ResultCod = DTOTask.ResultCods.DoneWithErrors
    End Sub

    Public Sub AddException(ex As System.Exception)
        AddException(ex.Message)
    End Sub

    Public Sub AddExceptions(exs As IEnumerable(Of System.Exception))
        For Each ex In exs
            AddException(ex.Message)
        Next
    End Sub

    Public Sub DoneWithErrors()
        _ResultCod = DTOTask.ResultCods.DoneWithErrors
    End Sub

    Public Sub DoneWithErrors(stringFormatMsg As String, ParamArray stringFormatValues() As String)
        DoneWithErrors()
        If stringFormatValues.Length > 0 Then
            _Msg = String.Format(stringFormatMsg, stringFormatValues)
        Else
            _Msg = stringFormatMsg
        End If
    End Sub

    Public Function ResultReport() As String
        Dim sb As New Text.StringBuilder
        sb.AppendLine("Cod: " & _ResultCod.ToString())
        sb.AppendLine("Msg: " & _Msg)
        For Each ex In _Exceptions
            sb.AppendLine(ex.Message)
        Next
        Dim retval As String = sb.ToString
        Return retval
    End Function



    Public Class Exception
        Property Message As String

        Public Sub New(sMessage As String)
            MyBase.New
            _Message = sMessage
        End Sub
    End Class

End Class

File: matiasmasso/Mat.NET/DTO/DTOTaskResult.cs
using System;
using System.Collections.Generic;

namespace DTO
{
    public class DTOTaskResult
    {
        public DTOTask.ResultCods ResultCod { get; set; }
        public string Msg { get; set; }
        public List<Exception> Exceptions { get; set; }

        public DTOTaskResult() : base()
        {
            Exceptions = new List<Exception>();
        }

        public static DTOTaskResult Factory(DTOTask.ResultCods oResultCod, string msg = "", List<Exception> exs = null)
        {
            DTOTaskResult retval = new DTOTaskResult();
            retval.ResultCod = oResultCod;
            retval.Msg = msg;
            if (exs != null)
                retval.AddExceptions(exs);
            return retval;
        }

        public bool Success()
        {
            bool retval = false;
            switch (ResultCod)
            {
                case DTOTask.ResultCods.Success:
                case DTOTask.ResultCods.Empty:
                    {
                        retval = true;
                        break;
                    }
            }
            return retval;
        }

        public static DTOTaskResult SuccessResult(string msg = "")
        {
            DTOTaskResult retval = new DTOTaskResult();
            retval.ResultCod = DTOTask.ResultCods.Success;
            retval.Msg = msg;
            return retval;
        }

        public static DTOTaskResult FailResult(List<System.Exception> exs, string msg = "")
        {
            DTOTaskResult retval = new DTOTaskResult();
            {
                var withBlock = retval;
                withBlock.ResultCod = DTOTask.ResultCods.Failed;
                withBlock.AddExceptions(exs);
                withBlock.Msg = msg;
            }
            return retval;
        }

        public void Fail()
        {
            ResultCod = DTOTask.ResultCods.Failed;
        }

        public void Fail(string stringFormatMsg, params string[] stringFormatValues)
        {
            Fail();
            if (stringFormatValues.Length > 0)
                Msg = string.Format(stringFormatMsg, stringFormatValues);
            else
                Msg = stringFormatMsg;
        }

        public void Fail(System.Exception ex, string stringFormatMsg, params string[] stringFormatValues)
        {
            Fail(stringFormatMsg, stringFormatValues);
            AddException(ex);
        }

        public void Fail(IEnumerable<System.Exception> exs, string stringFormatMsg, params string[] stringFormatValues)
        {
            Fail(stringFormatMsg, stringFormatValues);
            AddExceptions(exs);
        }

        public List<System.Exception> SystemExceptions()
        {
            List<System.Exception> retval = new List<System.Exception>();
            foreach (Exception ex in Exceptions)
                retval.Add(new System.Exception(ex.Message));
            return retval;
        }




        public void Succeed()
        {
            ResultCod = DTOTask.ResultCods.Success;
        }

        public void Succeed(string stringFormatMsg, params string[] stringFormatValues)
        {
            Succeed();
            if (stringFormatValues.Length > 0)
                Msg = string.Format(stringFormatMsg, stringFormatValues);
            else
                Msg = stringFormatMsg;
        }

        public void Empty()
        {
            ResultCod = DTOTask.ResultCods.Empty;
        }

        public void Empty(string stringFormatMsg, params string[] stringFormatValues)
        {
            Empty();
            if (stringFormatValues.Length > 0)
                Msg = string.Format(stringFormatMsg, stringFormatValues);
            else
                Msg = stringFormatMsg;
        }

        public void AddException(string stringFormatMsg, params string[] stringFormatValues)
        {
            Exception ex = null;
            if (stringFormatValues.Length > 0)
                ex = new Exception(string.Format(stringFormatMsg, stringFormatValues));
            else
                ex = new Exception(stringFormatMsg);
            Exceptions.Add(ex);
            if (ResultCod == DTOTask.ResultCods.Success)
                ResultCod = DTOTask.ResultCods.DoneWithErrors;
        }

        public void AddException(System.Exception ex)
        {
            AddException(ex.Message);
        }

        public void AddExceptions(IEnumerable<System.Exception> exs)
        {
            foreach (var ex in exs)
                AddException(ex.Message);
        }

        public void DoneWithErrors()
        {
            ResultCod = DTOTask.ResultCods.DoneWithErrors;
        }

        public void DoneWithErrors(string stringFormatMsg, params string[] stringFormatValues)
        {
            DoneWithErrors();
            if (stringFormatValues.Length > 0)
                Msg = string.Format(stringFormatMsg, stringFormatValues);
            else
                Msg = stringFormatMsg;
        }

        public string ResultReport()
        {
            System.Text.StringBuilder sb = new System.Text.StringBuilder();
            sb.AppendLine("Cod: " + ResultCod.ToString());
            sb.AppendLine("Msg: " + Msg);
            foreach (var ex in Exceptions)
                sb.AppendLine(ex.Message);
            string retval = sb.ToString();
            return retval;
        }


    }
}
